url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/DNSMessage.java

```diff
@@ -0,0 +1,1133 @@
+        if (optRecord == null) return null;
+        edns = new EDNS(optRecord);
+        return edns;
+    }
+
+    @SuppressWarnings("unchecked")
+    public Record<OPT> getOptPseudoRecord() {
+        if (optRrPosition == -1) return null;
+        return (Record<OPT>) additionalSection.get(optRrPosition);
+    }
+
+    /**
+     * Check if the EDNS DO (DNSSEC OK) flag is set.
+     *
+     * @return true if the DO flag is set.
+     */
+    public boolean isDnssecOk() {
+        EDNS edns = getEdns();
+        if (edns == null)
+            return false;
+
+        return edns.dnssecOk;
+    }
+
+    private String toStringCache;
+
+    @Override
+    public String toString() {
+        if (toStringCache != null) return toStringCache;
+
+        StringBuilder sb = new StringBuilder("DNSMessage")
+                .append('(').append(id).append(' ')
+                .append(opcode).append(' ')
+                .append(responseCode).append(' ');
+        if (qr) {
+            sb.append("resp[qr=1]");
+        } else {
+            sb.append("query[qr=0]");
+        }
+        if (authoritativeAnswer) sb.append(" aa");
+        if (truncated) sb.append(" tr");
+        if (recursionDesired) sb.append(" rd");
+        if (recursionAvailable) sb.append(" ra");
+        if (authenticData) sb.append(" ad");
+        if (checkingDisabled) sb.append(" cd");
+        sb.append(")\n");
+        if (questions != null) {
+            for (Question question : questions) {
+                sb.append("[Q: ").append(question).append("]\n");
+            }
+        }
+        if (answerSection != null) {
+            for (Record<? extends Data> record : answerSection) {
+                sb.append("[A: ").append(record).append("]\n");
+            }
+        }
+        if (authoritySection != null) {
+            for (Record<? extends Data> record : authoritySection) {
+                sb.append("[N: ").append(record).append("]\n");
+            }
+        }
+        if (additionalSection != null) {
+            for (Record<? extends Data> record : additionalSection) {
+                sb.append("[X: ");
+                EDNS edns = EDNS.fromRecord(record);
+                if (edns != null) {
+                    sb.append(edns.toString());
+                } else {
+                    sb.append(record);
+                }
+                sb.append("]\n");
+            }
+        }
+
+        // Strip trailing newline.
+        if (sb.charAt(sb.length() - 1) == '\n') {
+            sb.setLength(sb.length() - 1);
+        }
+
+        toStringCache = sb.toString();
+        return toStringCache;
+    }
+
+    private String terminalOutputCache;
+
+    /**
+     * Format the DNSMessage object in a way suitable for terminal output.
+     * The format is loosely based on the output provided by {@code dig}.
+     *
+     * @return This message as a String suitable for terminal output.
+     */
+    public String asTerminalOutput() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 22. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 725-731, lines 750-761 into separate methods.